### ISSUE TITLE
W3C WebDriver protocol support

### ISSUE THREAD

**Issue by OndraM** [php-webdriver/php-webdriver#469]
This is an issue to plan and track progress of implementing [W3C WebDriver](https://w3c.github.io/webdriver/webdriver-spec.html) protocol into php-webdriver.

Currently, only legacy [JsonWire](https://github.com/SeleniumHQ/selenium/wiki/JsonWireProtocol) (aka OSS) protocol is supported by php-webdriver. Selenium server 3.x translates the protocol if needed for the remote end (browser). However not everything is translated (like `actions()`), making some of the features not working with remote ends supporting only the W3C protocol (which is currently only GeckoDriver for Firefox 48+).

Since Selenium 3.5 the translation to old protocol is disabled by default. **It still could be reenabled** by starting the server like `java -jar selenium-server-standalone-3.5.3.jar -enablePassThrough false`, but this is only partial and temporary solution/workaround.

For Docker 🐳  users: use SE_OPTS environment variable to pass the enablePassThrough option to Selenium server. Ie. add `-e "SE_OPTS=-enablePassThrough false"` to `docker run` command.

## The work would require eg.:
- add some more functional tests, to make sure we don't breake the old protocol while doing the necessary refactoring
- prepare handshake to detect whether the remote end is speaking JsonWire or W3C dialect (we need this to support the old browsers etc.)
- define common interface for Json and W3C protocol
- convert current implementation to use this interface
- implement adapter for the W3C protocol
- implement GeckodriverService or something like that to start geckodriver locally
...

## Resources
- [JsonWireProtocol vs. W3C WebDriver](https://github.com/facebook/php-webdriver/wiki/JsonWireProtocol-vs.-W3C-WebDriver)
- [ProtocolHandshake](https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/remote/ProtocolHandshake.java) in java bindings
- [Handshake and capabilities resolving](https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/remote/webdriver.py#L252) in python bindings

**Comment by OndraM** [comment]
I started working on https://github.com/facebook/php-webdriver/wiki/JsonWireProtocol-vs.-W3C-WebDriver . I'd like to summarize there the key protocol differences before we start actually implementing them in the code. 

Any help with this overview document will be much appreciated!

**Comment by paulbriton** [comment]
Spec also contains changes about the JSON returned when finding an element. 
The key `ELEMENT` is no longer used for referencing an element so we have an undefined index in `RemoteWebDriver.php` line 178. See [this issue](https://github.com/SeleniumHQ/selenium/issues/4800) on selenium for more details. 

Here is an example of the JSON returned by IEDriverServer (which now implements the spec in its latest version, 3.6):
`{"state":"success","sessionId":null,"hCode":1164912850,"value":{"element-6066-11e4-a52e-4f735466cecf":"a1c4646c-8300-49f2-84b3-182911dc3eb6"},"class":"org.openqa.selenium.remote.Response","status":0}`

**Comment by davidmintz** [comment]
It would be great if the docs were updated to warn people about this and potentially save them a couple hours of pain. Are you interested in a PR to this effect?

**Comment by OndraM** [comment]
@davidmintz I've recently added note about passThourghMode to readme. However we can do more, for sure :). What is on your mind?

I'm also hoping to have some time for the issue itself during November...

**Comment by davidmintz** [comment]
@OndraM  sorry, I guess I missed that in the README, but I read your discussion about  ` -enablePassThrough false`, and tried it, but I am still exploding when I try to `sendKeys()`, with 
`Facebook\WebDriver\Exception\UnrecognizedExceptionException: Expected [object Undefined] undefined to be a string`  and I 'm unclear whether this rooted in the same issue as the undefined 'ELEMENT" index, or something else.

I do know it's making me crazy and I would love to help, just don't know if I have what it takes in terms of knowledge.

At the moment the only workaround that works for me is to use a very dated toolchain:  Selenium 2.44 with Firefox 35. With the current versions of everything plus Chrome, I get "chrome is unreachable" and I've been all over the map trying to solve that one.

**Comment by alexmnv** [comment]
Note for Docker users: Selenium parameters are passed through `SE_OPTS` env variable. i.e, you need to add `-e "SE_OPTS=-enablePassThrough false"` to `docker run` command

**Comment by paulbriton** [comment]
Hi @OndraM ! Happy new year ;)

Any news on this issue ? I've found this [Github repo](https://github.com/instaclick/php-webdriver) which is based on an old version of this project and which implements the W3C protocol (at least it says so). Maybe it could help !

**Comment by Roobix13** [comment]
Anyone can help me out, on how i can check my selenium version ?

**Comment by Mennims** [comment]
@Roobix13 I don't think this is the correct place to ask this. If you want to check your Selenium version use this syntax:
`java -jar "selenium server file name" --version`
If you need help use
`java -jar "selenium server file name" --help`.

You can also obtain status information including version number from the server as a json variable by navigating to "http://<seleniumserverip>:<port>/wd/hub/status/".

**Comment by Roobix13** [comment]
@Mennims I am sorry, i am new to the website and github. Thank you very much nevertheless.

**Comment by plista-mp** [comment]
For selenium-server-standalone version > 3.9.1: the option **enablePassThrough** doesn't work, just switch to version 3.8.1 to use this workaround for firefox

**Comment by OndraM** [comment]
Hi everyone, I'm aware of the situation, including the fact you cannot effectively use Selenium 3.9.0+ due to removal of the `enablePassThrough` - you must stick with the previous version for the time being. Sorry for this!

I also hope I will be able to share some more information about W3C protocol support quite soon. I know the missing support may be limiting for you now (but also workaround exists - see previous messages), but please hang tight :).

**Comment by OndraM** [comment]
Sorry, my previous statement was not accurate - you only cannot use Selenium 3.9 with remote ends supporting only the new W3C protocol (especially Firefox), because Selenium Server now won't do the translation to the new protocol. But those remote ends, which sill supports the old protocol (like Chrome) still DO work perfectly.

**Comment by sundaramrajendran** [comment]
Hi @OndraM ,

Do we have any ETA for this task?

**Comment by Joe33333** [comment]
Hi everyone,
I recently downloaded v 3.12 and the problems still persist (with Firefox). 
Can someone confirm that for me? Many thanks!

**Comment by GregJohnStewart** [comment]
yet another bump, pretty annoying issue. For anyone wanting a quick link to a version of selenium that works: https://selenium-release.storage.googleapis.com/index.html?path=3.8/

**Comment by schmunk42** [comment]
Latest version which works for me is still `2.53.1` - I get different errors with `3.8` (when filling a field) or `3.13` (can't open any page).

**Comment by williamdes** [comment]
From : https://www.w3.org/TR/webdriver/

Below you will find:
- List of W3 commands
- List of commands by constant name with url
- List of constants

:construction: **Work in progress** :construction:

Implemented | Method | URI Template | Command
-- | -- | -- | --
:stop_sign:  | POST | /session | New Session
:stop_sign: | DELETE | /session/{session id} | Delete Session
:stop_sign: | GET | /status | Status
:stop_sign: | GET | /session/{session id}/timeouts | Get Timeouts
:stop_sign: | POST | /session/{session id}/timeouts | Set Timeouts
:stop_sign: | POST | /session/{session id}/url | Navigate To
:stop_sign: | GET | /session/{session id}/url | Get Current URL
:stop_sign: | POST | /session/{session id}/back | Back
:stop_sign: | POST | /session/{session id}/forward | Forward
:stop_sign: | POST | /session/{session id}/refresh | Refresh
:stop_sign: | GET | /session/{session id}/title | Get Title
:stop_sign: | GET | /session/{session id}/window | Get Window Handle
:stop_sign: | DELETE | /session/{session id}/window | Close Window
:stop_sign: | POST | /session/{session id}/window | Switch To Window
:stop_sign: | GET | /session/{session id}/window/handles | Get Window Handles
:stop_sign: | POST | /session/{session id}/frame | Switch To Frame
:stop_sign: | POST | /session/{session id}/frame/parent | Switch To Parent Frame
:stop_sign: | GET | /session/{session id}/window/rect | Get Window Rect
:stop_sign: | POST | /session/{session id}/window/rect | Set Window Rect
:stop_sign: | POST | /session/{session id}/window/maximize | Maximize Window
:stop_sign: | POST | /session/{session id}/window/minimize | Minimize Window
:stop_sign: | POST | /session/{session id}/window/fullscreen | Fullscreen Window
:stop_sign: | GET | /session/{session id}/element/active | Get Active Element
:stop_sign: | POST | /session/{session id}/element | Find Element
:stop_sign: | POST | /session/{session id}/elements | Find Elements
:stop_sign: | POST | /session/{session id}/element/{element id}/element | Find Element From Element
:stop_sign: | POST | /session/{session id}/element/{element id}/elements | Find Elements From Element
:stop_sign: | GET | /session/{session id}/element/{element id}/selected | Is Element Selected
:stop_sign: | GET | /session/{session id}/element/{element id}/attribute/{name} | Get Element Attribute
:stop_sign: | GET | /session/{session id}/element/{element id}/property/{name} | Get Element Property
:stop_sign: | GET | /session/{session id}/element/{element id}/css/{property name} | Get Element CSS Value
:stop_sign: | GET | /session/{session id}/element/{element id}/text | Get Element Text
:stop_sign: | GET | /session/{session id}/element/{element id}/name | Get Element Tag Name
:stop_sign: | GET | /session/{session id}/element/{element id}/rect | Get Element Rect
:stop_sign: | GET | /session/{session id}/element/{element id}/enabled | Is Element Enabled
:stop_sign: | POST | /session/{session id}/element/{element id}/click | Element Click
:stop_sign: | POST | /session/{session id}/element/{element id}/clear | Element Clear
:stop_sign: | POST | /session/{session id}/element/{element id}/value | Element Send Keys
:stop_sign: | GET | /session/{session id}/source | Get Page Source
:stop_sign: | POST | /session/{session id}/execute/sync | Execute Script
:stop_sign: | POST | /session/{session id}/execute/async | Execute Async Script
:stop_sign: | GET | /session/{session id}/cookie | Get All Cookies
:stop_sign: | GET | /session/{session id}/cookie/{name} | Get Named Cookie
:stop_sign: | POST | /session/{session id}/cookie | Add Cookie
:stop_sign: | DELETE | /session/{session id}/cookie/{name} | Delete Cookie
:stop_sign: | DELETE | /session/{session id)/cookie | Delete All Cookies
:stop_sign: | POST | /session/{session id}/actions | Perform Actions
:stop_sign: | DELETE | /session/{session id}/actions | Release Actions
:stop_sign: | POST | /session/{session id}/alert/dismiss | Dismiss Alert
:stop_sign: | POST | /session/{session id}/alert/accept | Accept Alert
:stop_sign: | GET | /session/{session id}/alert/text | Get Alert Text
:stop_sign: | POST | /session/{session id}/alert/text | Send Alert Text
:stop_sign: | GET | /session/{session id}/screenshot | Take Screenshot
:stop_sign: | GET | /session/{session id}/element/{element id}/screenshot | Take Element Screenshot



W3 compatible | Method | URI Template | Constant
-- | -- | -- | --
:stop_sign: | POST | /session/{session id}/accept_alert | ACCEPT_ALERT
:stop_sign: | POST | /session/{session id}/cookie | ADD_COOKIE
:stop_sign: | POST | /session/{session id}/element/{element id}/clear | CLEAR_ELEMENT
:stop_sign: | POST | /session/{session id}/element/{element id}/click | CLICK_ELEMENT
:stop_sign: | DELETE | /session/{session id}/window | CLOSE
:stop_sign: | DELETE | /session/{session id}/cookie | DELETE_ALL_COOKIES
:stop_sign: | DELETE | /session/{session id}/cookie/:name | DELETE_COOKIE
:stop_sign: | POST | /session/{session id}/dismiss_alert | DISMISS_ALERT
:stop_sign: | GET | /session/{session id}/element/{element id}/equals/:other | ELEMENT_EQUALS
:stop_sign: | POST | /session/{session id}/element/{element id}/element | FIND_CHILD_ELEMENT
:stop_sign: | POST | /session/{session id}/element/{element id}/elements | FIND_CHILD_ELEMENTS
:stop_sign: | POST | /session/{session id}/execute | EXECUTE_SCRIPT
:stop_sign: | POST | /session/{session id}/execute_async | EXECUTE_ASYNC_SCRIPT
:stop_sign: | POST | /session/{session id}/element | FIND_ELEMENT
:stop_sign: | POST | /session/{session id}/elements | FIND_ELEMENTS
:stop_sign: | POST | /session/{session id}/frame | SWITCH_TO_FRAME
:stop_sign: | POST | /session/{session id}/window | SWITCH_TO_WINDOW
:stop_sign: | POST | /session/{session id}/url | GET
:stop_sign: | POST | /session/{session id}/element/active | GET_ACTIVE_ELEMENT
:stop_sign: | GET | /session/{session id}/alert_text | GET_ALERT_TEXT
:stop_sign: | GET | /session/{session id}/cookie | GET_ALL_COOKIES
:stop_sign: | GET | /sessions | GET_ALL_SESSIONS
:stop_sign: | GET | /session/{session id}/log/types | GET_AVAILABLE_LOG_TYPES
:stop_sign: | GET | /session/{session id}/url | GET_CURRENT_URL
:stop_sign: | GET | /session/{session id}/window_handle | GET_CURRENT_WINDOW_HANDLE
:stop_sign: | GET | /session/{session id}/element/{element id}/attribute/:name | GET_ELEMENT_ATTRIBUTE
:stop_sign: | GET | /session/{session id}/element/{element id}/css/:propertyName | GET_ELEMENT_VALUE_OF_CSS_PROPERTY
:stop_sign: | GET | /session/{session id}/element/{element id}/location | GET_ELEMENT_LOCATION
:stop_sign: | GET | /session/{session id}/element/{element id}/location_in_view | GET_ELEMENT_LOCATION_ONCE_SCROLLED_INTO_VIEW
:stop_sign: | GET | /session/{session id}/element/{element id}/size | GET_ELEMENT_SIZE
:stop_sign: | GET | /session/{session id}/element/{element id}/name | GET_ELEMENT_TAG_NAME
:stop_sign: | GET | /session/{session id}/element/{element id}/text | GET_ELEMENT_TEXT
:stop_sign: | POST | /session/{session id}/log | GET_LOG
:stop_sign: | GET | /session/{session id}/source | GET_PAGE_SOURCE
:stop_sign: | GET | /session/{session id}/orientation | GET_SCREEN_ORIENTATION
:stop_sign: | GET | /session/{session id} | GET_CAPABILITIES
:stop_sign: | GET | /session/{session id}/title | GET_TITLE
:stop_sign: | GET | /session/{session id}/window_handles | GET_WINDOW_HANDLES
:stop_sign: | GET | /session/{session id}/window/:windowHandle/position | GET_WINDOW_POSITION
:stop_sign: | GET | /session/{session id}/window/:windowHandle/size | GET_WINDOW_SIZE
:stop_sign: | POST | /session/{session id}/back | GO_BACK
:stop_sign: | POST | /session/{session id}/forward | GO_FORWARD
:stop_sign: | GET | /session/{session id}/element/{element id}/displayed | IS_ELEMENT_DISPLAYED
:stop_sign: | GET | /session/{session id}/element/{element id}/enabled | IS_ELEMENT_ENABLED
:stop_sign: | GET | /session/{session id}/element/{element id}/selected | IS_ELEMENT_SELECTED
:stop_sign: | POST | /session/{session id}/window/:windowHandle/maximize | MAXIMIZE_WINDOW
:stop_sign: | POST | /session/{session id}/buttondown | MOUSE_DOWN
:stop_sign: | POST | /session/{session id}/buttonup | MOUSE_UP
:stop_sign: | POST | /session/{session id}/click | CLICK
:stop_sign: | POST | /session/{session id}/doubleclick | DOUBLE_CLICK
:stop_sign: | POST | /session/{session id}/moveto | MOVE_TO
:stop_sign: | POST | /session | NEW_SESSION
:stop_sign: | DELETE | /session/{session id} | QUIT
:stop_sign: | POST | /session/{session id}/refresh | REFRESH
:stop_sign: | POST | /session/{session id}/file | UPLOAD_FILE // undocumented
:stop_sign: | POST | /session/{session id}/keys | SEND_KEYS_TO_ACTIVE_ELEMENT
:stop_sign: | POST | /session/{session id}/alert_text | SET_ALERT_VALUE
:stop_sign: | POST | /session/{session id}/element/{element id}/value | SEND_KEYS_TO_ELEMENT
:stop_sign: | POST | /session/{session id}/timeouts/implicit_wait | IMPLICITLY_WAIT
:stop_sign: | POST | /session/{session id}/orientation | SET_SCREEN_ORIENTATION
:stop_sign: | POST | /session/{session id}/timeouts | SET_TIMEOUT
:stop_sign: | POST | /session/{session id}/timeouts/async_script | SET_SCRIPT_TIMEOUT
:stop_sign: | POST | /session/{session id}/window/:windowHandle/position | SET_WINDOW_POSITION
:stop_sign: | POST | /session/{session id}/window/:windowHandle/size | SET_WINDOW_SIZE
:stop_sign: | POST | /session/{session id}/element/{element id}/submit | SUBMIT_ELEMENT
:stop_sign: | GET | /session/{session id}/screenshot | SCREENSHOT
:stop_sign: | POST | /session/{session id}/touch/click | TOUCH_SINGLE_TAP
:stop_sign: | POST | /session/{session id}/touch/down | TOUCH_DOWN
:stop_sign: | POST | /session/{session id}/touch/doubleclick | TOUCH_DOUBLE_TAP
:stop_sign: | POST | /session/{session id}/touch/flick | TOUCH_FLICK
:stop_sign: | POST | /session/{session id}/touch/longclick | TOUCH_LONG_PRESS
:stop_sign: | POST | /session/{session id}/touch/move | TOUCH_MOVE
:stop_sign: | POST | /session/{session id}/touch/scroll | TOUCH_SCROLL
:stop_sign: | POST | /session/{session id}/touch/up | TOUCH_UP


Name | Real Name
-- | --
GET_ALL_SESSIONS | getAllSessions
GET_CAPABILITIES | getCapabilities
NEW_SESSION | newSession
STATUS | status
CLOSE | close
QUIT | quit
GET | get
GO_BACK | goBack
GO_FORWARD | goForward
REFRESH | refresh
ADD_COOKIE | addCookie
GET_ALL_COOKIES | getCookies
DELETE_COOKIE | deleteCookie
DELETE_ALL_COOKIES | deleteAllCookies
FIND_ELEMENT | findElement
FIND_ELEMENTS | findElements
FIND_CHILD_ELEMENT | findChildElement
FIND_CHILD_ELEMENTS | findChildElements
CLEAR_ELEMENT | clearElement
CLICK_ELEMENT | clickElement
SEND_KEYS_TO_ELEMENT | sendKeysToElement
SEND_KEYS_TO_ACTIVE_ELEMENT | sendKeysToActiveElement
SUBMIT_ELEMENT | submitElement
UPLOAD_FILE | uploadFile
GET_CURRENT_WINDOW_HANDLE | getCurrentWindowHandle
GET_WINDOW_HANDLES | getWindowHandles
GET_CURRENT_CONTEXT_HANDLE | getCurrentContextHandle
GET_CONTEXT_HANDLES | getContextHandles
SWITCH_TO_WINDOW | switchToWindow
SWITCH_TO_CONTEXT | switchToContext
SWITCH_TO_FRAME | switchToFrame
SWITCH_TO_PARENT_FRAME | switchToParentFrame
GET_ACTIVE_ELEMENT | getActiveElement
GET_CURRENT_URL | getCurrentUrl
GET_PAGE_SOURCE | getPageSource
GET_TITLE | getTitle
EXECUTE_SCRIPT | executeScript
EXECUTE_ASYNC_SCRIPT | executeAsyncScript
GET_ELEMENT_TEXT | getElementText
GET_ELEMENT_TAG_NAME | getElementTagName
IS_ELEMENT_SELECTED | isElementSelected
IS_ELEMENT_ENABLED | isElementEnabled
IS_ELEMENT_DISPLAYED | isElementDisplayed
GET_ELEMENT_LOCATION | getElementLocation
GET_ELEMENT_LOCATION_ONCE_SCROLLED_INTO_VIEW | getElementLocationOnceScrolledIntoView
GET_ELEMENT_SIZE | getElementSize
GET_ELEMENT_ATTRIBUTE | getElementAttribute
GET_ELEMENT_VALUE_OF_CSS_PROPERTY | getElementValueOfCssProperty
ELEMENT_EQUALS | elementEquals
SCREENSHOT | screenshot
ACCEPT_ALERT | acceptAlert
DISMISS_ALERT | dismissAlert
GET_ALERT_TEXT | getAlertText
SET_ALERT_VALUE | setAlertValue
SET_TIMEOUT | setTimeout
IMPLICITLY_WAIT | implicitlyWait
SET_SCRIPT_TIMEOUT | setScriptTimeout
EXECUTE_SQL | executeSQL
GET_LOCATION | getLocation
SET_LOCATION | setLocation
GET_APP_CACHE | getAppCache
GET_APP_CACHE_STATUS | getStatus
CLEAR_APP_CACHE | clearAppCache
IS_BROWSER_ONLINE | isBrowserOnline
SET_BROWSER_ONLINE | setBrowserOnline
GET_LOCAL_STORAGE_ITEM | getLocalStorageItem
GET_LOCAL_STORAGE_KEYS | getLocalStorageKeys
SET_LOCAL_STORAGE_ITEM | setLocalStorageItem
REMOVE_LOCAL_STORAGE_ITEM | removeLocalStorageItem
CLEAR_LOCAL_STORAGE | clearLocalStorage
GET_LOCAL_STORAGE_SIZE | getLocalStorageSize
GET_SESSION_STORAGE_ITEM | getSessionStorageItem
GET_SESSION_STORAGE_KEYS | getSessionStorageKey
SET_SESSION_STORAGE_ITEM | setSessionStorageItem
REMOVE_SESSION_STORAGE_ITEM | removeSessionStorageItem
CLEAR_SESSION_STORAGE | clearSessionStorage
GET_SESSION_STORAGE_SIZE | getSessionStorageSize
SET_SCREEN_ORIENTATION | setScreenOrientation
GET_SCREEN_ORIENTATION | getScreenOrientation
CLICK | mouseClick
DOUBLE_CLICK | mouseDoubleClick
MOUSE_DOWN | mouseButtonDown
MOUSE_UP | mouseButtonUp
MOVE_TO | mouseMoveTo
IME_GET_AVAILABLE_ENGINES | imeGetAvailableEngines
IME_GET_ACTIVE_ENGINE | imeGetActiveEngine
IME_IS_ACTIVATED | imeIsActivated
IME_DEACTIVATE | imeDeactivate
IME_ACTIVATE_ENGINE | imeActivateEngine
TOUCH_SINGLE_TAP | touchSingleTap
TOUCH_DOWN | touchDown
TOUCH_UP | touchUp
TOUCH_MOVE | touchMove
TOUCH_SCROLL | touchScroll
TOUCH_DOUBLE_TAP | touchDoubleTap
TOUCH_LONG_PRESS | touchLongPress
TOUCH_FLICK | touchFlick
SET_WINDOW_SIZE | setWindowSize
SET_WINDOW_POSITION | setWindowPosition
GET_WINDOW_SIZE | getWindowSize
GET_WINDOW_POSITION | getWindowPosition
MAXIMIZE_WINDOW | maximizeWindow
GET_AVAILABLE_LOG_TYPES | getAvailableLogTypes
GET_LOG | getLog
GET_SESSION_LOGS | getSessionLogs
GET_NETWORK_CONNECTION | getNetworkConnection
SET_NETWORK_CONNECTION | setNetworkConnection

**Comment by mpierzchalski** [comment]
Hi, here is my PR https://github.com/facebook/php-webdriver/pull/599 with W3C support. It's not perfect but works with Firefox 60.x (https://travis-ci.org/facebook/php-webdriver/jobs/430392005) and latest Chrome. I don't know why it fails in SauceLabs, but fell free to try it and test it on your own.

**Comment by kjnanda** [comment]
This issue still persists for selenium 3.14. Getting the same console error while running the script.

**Comment by jonah** [comment]
These changes would be very useful to me to get browser testing working in selenium/os x.  I've checked out this pull request and it solves the w3c/enablePassthrough errors I was experiencing.   Is there anything I can do to contribute?  Is there a fork of php-webdriver that someone has created that has these features that I can install with composer in the meantime?

**Comment by vkravc** [comment]
Still not working firefox with Selenium 3.141.59

**Comment by matvey88** [comment]
Stopped working php-webdriver on Safari 
Says [PHPUnit\Framework\Exception] Undefined index: ELEMENT on command $I->fillField $I->waitForText and $I->waitForElement

About a two weeks ago this commands working
Can you fix it?)
Thanks )

**Comment by andrewnicols** [comment]
Looks like they dropped the legacy ELEMENT index recently:
https://bugzilla.mozilla.org/show_bug.cgi?id=1400233

The `ELEMENT` index was a legacy index and dropped in FF 63.0.

**Comment by schmunk42** [comment]
@OndraM Is this project still alive?

> Is there an alternative project which has implemented the spec in its current form?

I am asking because we're currently relying on the webdriver in conjunction with Codeception (PHP) and we need to make a decision if we should finally move over to CodeceptJS for acceptance tests.

**Comment by DanielRuf** [comment]
> Looks like they dropped the legacy ELEMENT index recently:
> [bugzilla.mozilla.org/show_bug.cgi?id=1400233](https://bugzilla.mozilla.org/show_bug.cgi?id=1400233)
> 
> The `ELEMENT` index was a legacy index and dropped in FF 63.0.

Same issue on IE11 with the IEDriver.

**Comment by SagarNaliyapara** [comment]
@OndraM 
I'm getting following error,

```
Facebook\WebDriver\Exception\UnknownCommandException: POST /session/0a263573c608002afc74fa6b59aeb0d0/element//clear
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:25:53'
System info: host: 'f4107bb89fa3', ip: '172.25.1.3', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-1013-aws', java.version: '1.8.0_212'
Driver info: driver.version: unknown
```

**Comment by DanielRuf** [comment]
@SagarNaliyapara this is not related to this issue. Please open a new issue.

**Comment by reinholdfuereder** [comment]
> 
> > Looks like they dropped the legacy ELEMENT index recently:
> > [bugzilla.mozilla.org/show_bug.cgi?id=1400233](https://bugzilla.mozilla.org/show_bug.cgi?id=1400233)
> > The `ELEMENT` index was a legacy index and dropped in FF 63.0.
> 
> Same issue on IE11 with the IEDriver.

I naively guess a similar (or the same?) problem with Chrome v75 (via https://github.com/SeleniumHQ/docker-selenium/releases/tag/3.141.59-palladium; the previous releases with Chrome and ChromeDriver v74 still worked), because while according to https://sites.google.com/a/chromium.org/chromedriver/downloads

> The most noticeable change is ChromeDriver now runs in W3C standard compliant mode by default

... sounds like everything should still work when switching to non-W3C mode (please especially note the `Detected dialect: OSS` output instead of default `Detected dialect: W3C`):

```
13:20:00.890 INFO [ActiveSessionFactory.apply] - Capabilities are: {
  "browserName": "chrome",
  "goog:chromeOptions": {
    "args": [
      "--headless",
      "--disable-gpu",
      "--window-size=1920x1200"
    ],
    "w3c": false
  }
}
13:20:00.890 INFO [ActiveSessionFactory.lambda$apply$11] - Matched factory org.openqa.selenium.grid.session.remote.ServicedSession$Factory (provider: org.openqa.selenium.chrome.ChromeDriverService)
Starting ChromeDriver 75.0.3770.8 (681f24ea911fe754973dda2fdc6d2a2e159dd300-refs/branch-heads/3770@{#40}) on port 22008
Only local connections are allowed.
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
[1559827200.907][SEVERE]: bind() failed: Cannot assign requested address (99)
13:20:01.113 INFO [ProtocolHandshake.createSession] - Detected dialect: OSS
13:20:01.161 INFO [RemoteSession$Factory.lambda$performHandshake$0] - Started new session 55234a7ec928b26e1dff81ea0b7cd803 (org.openqa.selenium.chrome.ChromeDriverService)
13:20:16.955 INFO [ActiveSessions$1.onStop] - Removing session 55234a7ec928b26e1dff81ea0b7cd803 (org.openqa.selenium.chrome.ChromeDriverService)
```
... the (Codeception v2.5.6) test then fails with `Undefined index: ELEMENT` (like in e.g. #492 or #576) => :-o or ;-(

**Comment by andrewnicols** [comment]
I am currently testing an updated version of this repo which merges the existing W3C branch, fixes chrome, mouse release, and a few other bits.

Not sure if we’ll be forking or just maintaining an alias repository but I’ll be pushing my changes in the next few days.

It does seem that this project and Instaclick are both dead, but if someone would announce officially it would really help in making some decisions for the rest of us.

**Comment by reinholdfuereder** [comment]
@andrewnicols thanks for your extremely fast and promising response => looking forward very much indeed...

**Comment by andrewnicols** [comment]
I’m not sure if we’ll be able to maintain a fork, but if we can get to a point where someone, or some group of people, can...

Note: chromedriver 75 only made w3c compliance mode on by default. You can still pass w3c: false as a required capability to get legacy behaviour I think.

**Comment by reinholdfuereder** [comment]
@andrewnicols Oops, my comment https://github.com/facebook/php-webdriver/issues/469#issuecomment-499497795 was not correct, sorry => see correction at the bottom

Ad "maintain a fork": yeah, I hope so; it looks as if there is enough demand and therefore potentially a group of interested devs around...

Ad "w3c: false for legacy mode" and **correction of my comment**: yes, that is what I actually tried, but it did not help me in my simple Codeception test for cookie policy (cookie policy banner ok button click hides the cookie policy banner):
* when using default W3C protocol, then the Codeception (v2.5.6) test fails with `Undefined index: ELEMENT` (like in e.g. #492 or #576), in `waitForElement()` operation
* when using legacy protocol, then the test succeeds for `waitForElement()` operation, which is followed by a seemingly also successful `click()` operation, but fails in the `waitForElementNotVisible()` operation by running into the default 10s timeout (when waiting for the hiding of the element due to the mouse click -- as if the click did not really happen or whatever)
  * frankly I am not sure why that happens, or if that could be actually another unrelated problem: it used to work ever since (that is at least since Chrome v70)

**Comment by andrewnicols** [comment]
Took me longer than I had hoped, but I've gotten to a point where _most_ things are working in https://github.com/andrewnicols/php-webdriver/tree/fixtheworld

I still have a list of things not yet working (GET_ELEMENT_LOCATION_ONCE_SCROLLED_INTO_VIEW is not available in W3C Webdriver for example) and I discovered that the window API has changed (now fixed). I need to  go through the complete list of commands and check them off against the spec.

**Comment by OndraM** [comment]
Hi everyone, sorry for not moving this issue forward in recent months :-(. 

About the current issue with Chrome 75 - it enables W3C protocol by default. Unfortunately just disabling it wasn't enough, as this change in Chromedriver introduced some bugs. However I was able to find a workaround for the Chromedriver bug - see #640 . Because this issue affects lots of people and other projects (eg. https://github.com/laravel/dusk/issues/651), the plan is to release this change ASAP as a new version.

And about future and full implementation of the W3C protocol. **This project is definitely not dead**. In next days, I will wrote down some ideas about future of the project and a ways how to get the W3C protocol implemented.

It's not secret this project requires some more maintainers to help with the W3C implementation and push it forward. I'm quite impressed by recent work of @andrewnicols on top of the #560 by @dunglas, great work! If you are interested, it would be great if I can include you and maybe others in this effort :pray: .

**Comment by dunglas** [comment]
I'm interested in working with @andrewnicols and you to merge W3C support as soon as possible. It's very important for Symfony / Panther. Let me know how I can help.
I'm also ok to help maintaining this lib.

**Comment by OndraM** [comment]
Immediate fix for Chrome 75 has been released as part of php-webdriver 1.7.0.

**Comment by andrewnicols** [comment]
Hi @OndraM,

I'm keen to help with this effort, and to share maintenance of the project.

Any chance you can share your plans and ideas for the future of the project?

**Comment by OneEyedSpaceFish** [comment]
Hey @OndraM ;

I'm also interested of getting involved in this project.
Let me know how we can help.

**Comment by andrewnicols** [comment]
Hi @OndraM, is there any chance that we can get the ball rolling again with this project?

We're all stuck in the dark ages testing with Firefox 47.0.1 and there are people like @dunglas (Symfony/Panther) and myself (Moodle) who would love to help get things moving.

**Comment by andrewnicols** [comment]
Hi @OndraM, pinging again. We really need to do something with this project to make it compatible with W3C.

**Comment by joaomoraes-webjump** [comment]
Hi @OndraM ping :dancing_women:

**Comment by OndraM** [comment]
Hi again, I'm so sorry for the delays :disappointed: and really unhappy with not pushing this forward recently, but my time resources are limited and I needed to spend lot of work on other job-stuff.

But I've prepared request for comments for php-webdriver 2.0 - see #657 , and I invite everyone interested to check it out and comment. So that once I'm in a short while back from upcoming vacation, we can finally start the hard work. Thanks everyone!

**Comment by OndraM** [comment]
### :loudspeaker: Current status update:

- **The experimental W3C protocol support (#560) was finally merged to community branch!** Thanks @dunglas and others involved! :100: :100: :100: 
- If remote browser is capable of W3C protocol, it will now be automatically used instead of the old OSS JsonWireProtocol. This means eg. Chromedriver will now *by default* use W3C protocol. (This could still be disabled if you need to - see https://github.com/facebook/php-webdriver/pull/674#issue-339395010 )
- This also means Geckodriver (ie. any new Firefox) could now be used! :fox_face: 
- Even though library code coverage is far from being ideal, we have passing Travis test builds for various browsers and protocol combinations: Latest Chrome with JsonWire protocol, with W3C protocol, via Selenium server, directly via chromedriver; Firefox 45 using JsonWire protocol, Firefox 67 using W3C protocol via Geckodriver and also via Selenium server... So lets hope we won't break much things :four_leaf_clover: .

What needs to be mentioned again - **the W3C protocol support is experimental** (and will remain experimental in upcoming php-webdriver 1.8). Some parts (like capabilities) are not yet implemented to fully conform W3C WebDriver protocol. However most of the stuff should work.

### What will be the process now:
- **Please test!** You can easily update to the development version for the upcoming 1.8 release by changing dependency in your composer.json like this: 
```json
    "require": {
        "facebook/webdriver": "^1.8@dev"
    }
```
and running `composer update`.
(Or you can just run `composer require facebook/webdriver:^1.8@dev`).
- **If you encounter any issues, please report them!**
- This experimental W3C support will be released as part of php-webdriver 1.8, what should hopefully happen soon (if nothing bad happens). 
- Further steps towards future php-webdriver 2.0 are discussed in #657 . Also full support (not experimental) of W3C WebDriver will be part of php-webdriver 2.0.

:fire: :fire: :fire:

**Comment by OndraM** [comment]
The `community` branch now contains further W3C protocol-related improvements:
- Capabilities are converted to W3C compatible ones
- Some window-realted commands (set size, maximize etc.) were added to those supported in W3C protocol mode

:bell: **Everyone please test the community branch and report issues!**
`composer require facebook/webdriver:^1.8@dev`

**Comment by sathyas123** [comment]
Hi @OndraM,

I tried using the community branch (facebook/webdriver:^1.8@dev) for running my tests on firefox latest version and I am able to open the firefox, load the page but the elements are not found. The same test works fine on chrome lastest version. Could you please help me in running the tests on latest firefox?

Platform: Linux
Firefox version: 71.0 (64-bit)
Selenium standalone server: 3.141.59
Gecko driver version: v0.24.0
PHP version: PHP 7.3.12 (cli) 

**behat.yml:**

 ```
 extensions:
    Behat\MinkExtension:
      default_session:  selenium2
      base_url: 'https://login.yahoo.com/'
      browser_name: firefox
      selenium2:
        capabilities:
          browser: firefox
          marionette: true
          version: ''
```

**Composer.json:**

 ```
"require": {
    "emuse/behat-html-formatter": "dev-master as 0.1.99",
    "behat/behat": "^3.4",
    "behat/mink-extension": "2.3.1",
    "behat/mink-selenium2-driver": "1.3.1",
    "phpunit/phpunit": "~7.1",
    "phpunit/phpunit-selenium": "~7.0",
    "monolog/monolog": "^1.22",
    "league/csv": "^8.2",
    "guzzlehttp/guzzle": "^6.3",
    "fzaninotto/faker": "@stable",
    "facebook/webdriver": "^1.8@dev",
    "php": "7.3.12"
  },
```

**Starting selenium standalone server and gecko driver:**

`java -jar -Dwebdriver.gecko.driver=geckodriver selenium-server-standalone-3.141.59.jar`


**Test Feature:**

```
  @testfirefox
  Scenario: Open yahoo mail and click forgot password
    When User launch Application
    And User click forgot password link

```
Step definition for User click forgot password link

```
  $this->wait(10);
  print_r($this->getSession()->getDriver()->click("//a[@id='mbr-forgot-link']"));
  $this->wait(10);

```

**Test Output Error**

`(Element not found with xpath, //a[@id='mbr-forgot-link'] )`

**Note:**

When I try to give 
print_r($this->getSession()->getPage()->getContent()); 
I am getting the content. 
But when I try to give 
print_r($this->getSession()->getDriver()->getHtml("//html")); 
I am getting the error (Element not found with xpath, //html )

**Selenium Log:**

>

```
 16:28:04.817 INFO [ActiveSessionFactory.apply] - Capabilities are: {
  "browser": "firefox",
  "browserName": "firefox",
  "ignoreZoomSetting": false,
  "marionette": true,
  "name": "Behat feature suite",
  "tags": [
    "LT00860",
    "PHP 7.3.12"
  ],
  "version": ""
}
16:28:04.818 INFO [ActiveSessionFactory.lambda$apply$11] - Matched factory org.openqa.selenium.grid.session.remote.ServicedSession$Factory (provider: org.openqa.selenium.firefox.GeckoDriverService)
1577118485007	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "-marionette" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileeS6Xv9"
1577118485605	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1577118485605	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1577118485605	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1577118485605	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1577118487770	Marionette	INFO	Listening on port 39591
16:28:07.880 INFO [ProtocolHandshake.createSession] - Detected dialect: W3C
16:28:07.931 INFO [RemoteSession$Factory.lambda$performHandshake$0] - Started new session 761450de-03e1-4b4a-a62c-8ae1144b57a5 (org.openqa.selenium.firefox.GeckoDriverService)
1577118498783	Marionette	INFO	Stopped listening on port 39591
16:28:20.497 INFO [ActiveSessions$1.onStop] - Removing session 761450de-03e1-4b4a-a62c-8ae1144b57a5 (org.openqa.selenium.firefox.GeckoDriverService)
```

**Comment by OndraM** [comment]
@sathyas123 Even though you have installed php-webdriver, from what I understand, you are using `mink-selenium2-driver` - that means you don't actually use this library in your tests, as Behat uses its own, outdated Selenium JsonWire protocol implementation, see https://github.com/minkphp/MinkSelenium2Driver/issues/293

However please report any further issues as a [new standalone ticket](https://github.com/facebook/php-webdriver/issues/new) here on Github, so that this thread don't get too flooded.

**Comment by andrewnicols** [comment]
@OndraM, do you think that we may be able to release 1.8.0 shortly? I think we're nearly up to full compliance with the W3C spec (just a couple of missing commands which mostly have pull requests). Do you have a timeframe in mind?

**Comment by OndraM** [comment]
No exact timeframe, but I've created [1.8.0 milestone](https://github.com/facebook/php-webdriver/milestone/2) to track reamining issues. There is just a few of them, so it should not take long. And I don't think we necessarily need *full* W3C compliance, this would be aim of 2.0. We just need "good enough" implementation in 1.8 :-).

**Comment by OndraM** [comment]
:loudspeaker: One more announcement for everyone watching this issue and waiting for W3C support.

As we move towards release 1.8.0, we successfully transferred the repository (as explained in #730) and it is now available under new name. So in order to get any future updates of the package, *you must update its name in require section of your composer.json file* like this:

```diff
"require": {
-    "facebook/webdriver": "(version you use)",
+    "php-webdriver/webdriver": "(version you use)",
}
```

**Comment by OndraM** [comment]
### :tada: Version 1.8.0 was released!

Thanks to all 14 community contributors who made this possible (in 120 commits)! :1st_place_medal: :heart: 

:new: Have look at [complete changelog](https://github.com/php-webdriver/php-webdriver/blob/master/CHANGELOG.md#180---2020-02-10).

Version 1.8.0 includes experimental support of W3C WebDriver protocol, and it will be used by default if your browser supports it (recommended is latest Chrome/Firefox). If you encounter any issues with the W3C protocol, you can still [explicitly disable the protocol](https://github.com/php-webdriver/php-webdriver/pull/674#issue-339395010) and use the legacy "JsonWire OSS protocol", but please also [report the issue](https://github.com/php-webdriver/php-webdriver/issues/new) here on GitHub.

Also don't forget to rename the library in your composer.json:

```diff
"require": {
-    "facebook/webdriver": "...",
+    "php-webdriver/webdriver": "^1.8.0",
}
```
**before** you run `composer update`.

Happy testing!
